btcec/schnorr/signature.go

```diff
@@ -0,0 +1,558 @@
+	// point.
+
+	// Step 1.
+	//
+	// Fail if m is not 32 bytes
+	if len(hash) != scalarSize {
+		str := fmt.Sprintf("wrong size for message (got %v, want %v)",
+			len(hash), scalarSize)
+		return signatureError(ecdsa_schnorr.ErrInvalidHashLen, str)
+	}
+
+	// Before we proceed, we want to ensure that the public key we're using
+	// for verification always has an even y-coordinate. So we'll serialize
+	// it, then parse it again to esure we only proceed with points that
+	// have an even y-coordinate.
+	pubKey, err := ParsePubKey(pub.SerializeCompressed()[1:])
+	if err != nil {
+		return err
+	}
+
+	// Step 2.
+	//
+	// Fail if Q is not a point on the curve
+	if !pubKey.IsOnCurve() {
+		str := "pubkey point is not on curve"
+		return signatureError(ecdsa_schnorr.ErrPubKeyNotOnCurve, str)
+	}
+
+	// Step 3.
+	//
+	// Fail if r >= p
+	//
+	// Note this is already handled by the fact r is a field element.
+
+	// Step 4.
+	//
+	// Fail if s >= n
+	//
+	// Note this is already handled by the fact s is a mod n scalar.
+
+	// Step 5.
+	//
+	// e = int(tagged_hash("BIP0340/challenge", bytes(r) || bytes(P) || M)) mod n.
+	var rBytes [32]byte
+	sig.r.PutBytesUnchecked(rBytes[:])
+	pBytes := pubKey.SerializeCompressed()
+
+	commitment := chainhash.TaggedHash(
+		tagHashChallenge, rBytes[:], pBytes[1:], hash,
+	)
+
+	var e btcec.ModNScalar
+	if overflow := e.SetBytes((*[32]byte)(commitment)); overflow != 0 {
+		str := "hash of (r || P || m) too big"
+		return signatureError(ecdsa_schnorr.ErrSchnorrHashValue, str)
+	}
+
+	// Negate e here so we can use AddNonConst below to subtract the s*G
+	// point from e*P.
+	e.Negate()
+
+	// Step 6.
+	//
+	// R = s*G - e*P
+	var P, R, sG, eP btcec.JacobianPoint
+	pubKey.AsJacobian(&P)
+	btcec.ScalarBaseMultNonConst(&sig.s, &sG)
+	btcec.ScalarMultNonConst(&e, &P, &eP)
+	btcec.AddNonConst(&sG, &eP, &R)
+
+	// Step 7.
+	//
+	// Fail if R is the point at infinity
+	if (R.X.IsZero() && R.Y.IsZero()) || R.Z.IsZero() {
+		str := "calculated R point is the point at infinity"
+		return signatureError(ecdsa_schnorr.ErrSigRNotOnCurve, str)
+	}
+
+	// Step 8.
+	//
+	// Fail if R.y is odd
+	//
+	// Note that R must be in affine coordinates for this check.
+	R.ToAffine()
+	if R.Y.IsOdd() {
+		str := "calculated R y-value is odd"
+		return signatureError(ecdsa_schnorr.ErrSigRYIsOdd, str)
+	}
+
+	// Step 9.
+	//
+	// Verified if R.x == r
+	//
+	// Note that R must be in affine coordinates for this check.
+	if !sig.r.Equals(&R.X) {
+		str := "calculated R point was not given R"
+		return signatureError(ecdsa_schnorr.ErrUnequalRValues, str)
+	}
+
+	// Step 10.
+	//
+	// Return success iff not failure occured before reachign this
+	return nil
+}
+
+// Verify returns whether or not the signature is valid for the provided hash
+// and secp256k1 public key.
+func (sig *Signature) Verify(hash []byte, pubKey *btcec.PublicKey) bool {
+	return schnorrVerify(sig, hash, pubKey) == nil
```

Comment:
@gcsfred2 @wpaulino alternatively, we can just serialize the public key here (which is essentially the same as the `lift_x` algo in the BIP). At that point our `schnorrVerify` is pure from the PoV of the BIP.

Reply:
Latest commit includes this change, hopefully it's less confusing now? The internal `schnorrVerify` now always accepts a serialized 32-byte-x-only public key. The caller doesn't need to serialize the key, as the main `Verify` function does. The end algorithm is the same, but the internal function is now more "pure" from the PoV of the BIP.